### PR TITLE
mdls table implementation

### PR DESCRIPTION
**Pull request by drakearonhalt**
Created mdls table which mimics the functionality of the mdls command in macOS. Table allows osquery to retrieve key/value pairs from spotlight metadata.

### CODE REVIEW

File: osquery/tables/system/darwin/mdls.mm [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+  MDItemRef mdi = MDItemCreate(kCFAllocatorDefault, cs);
+  CFTypeRef tr = MDItemCopyAttribute(mdi, CFSTR("kMDItemPath"));
+  CFArrayRef al = MDItemCopyAttributeNames(mdi);
+  CFDictionaryRef d = MDItemCopyAttributes(mdi, al);
+  for (int j = 0; j < CFArrayGetCount(al); ++j) {
+    CFTypeRef key = CFArrayGetValueAtIndex(al, j);
+    CFTypeRef value = CFDictionaryGetValue(d, key);
+    std::string rvalue = "null";
+
+    if (tr == nullptr) {
```
Comment (obelisk) [inline osquery/tables/system/darwin/mdls.mm]: Can you turn this into one statement with ||

File: osquery/tables/system/darwin/mdls.mm [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+#include "osquery/core/conversions.h"
+
+namespace osquery {
+namespace tables {
+
+std::string getStringOfValue(CFTypeRef value) {
+  std::string rvalue;
+
+  if (CFGetTypeID(value) == CFNumberGetTypeID()) {
+    rvalue = stringFromCFNumber((CFDataRef)value);
```
Comment (obelisk) [inline osquery/tables/system/darwin/mdls.mm]: Please avoid c style casts

File: osquery/tables/system/darwin/mdls.mm [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+  } else if (CFGetTypeID(value) == CFDateGetTypeID()) {
+    auto unix_time = CFDateGetAbsoluteTime((CFDateRef)value) +
```
Comment (obelisk) [inline osquery/tables/system/darwin/mdls.mm]: Please avoid c style casts

File: osquery/tables/system/darwin/mdls.mm [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                     kCFAbsoluteTimeIntervalSince1970;
+    rvalue = INTEGER(std::llround(unix_time));
+  } else if (CFGetTypeID(value) == CFArrayGetTypeID()) {
+    for (int i = 0; i < CFArrayGetCount((CFArrayRef)value); ++i) {
```
Comment (obelisk) [inline osquery/tables/system/darwin/mdls.mm]: Please avoid c style casts

File: osquery/tables/system/darwin/mdls.mm [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+      CFTypeRef b = CFArrayGetValueAtIndex((CFArrayRef)value, i);
```
Comment (obelisk) [inline osquery/tables/system/darwin/mdls.mm]: Please avoid c style casts

File: osquery/tables/system/darwin/mdls.mm [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+      if (i == 0) {
+        rvalue = getStringOfValue(b);
+      } else {
+        // beware of recursion....
+        rvalue = rvalue + "," + getStringOfValue(b);
+      }
+    }
+  } else if (CFGetTypeID(value) == CFBooleanGetTypeID()) {
+    if (CFBooleanGetValue((CFBooleanRef)value)) {
```
Comment (obelisk) [inline osquery/tables/system/darwin/mdls.mm]: Please avoid c style casts

File: osquery/tables/system/darwin/mdls.mm [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+      rvalue = "true";
+    } else {
+      rvalue = "false";
+    }
+  } else if (CFGetTypeID(value) == CFStringGetTypeID()) {
+    CFDataRef df = CFStringCreateExternalRepresentation(
+        kCFAllocatorDefault, (CFStringRef)value, kCFStringEncodingASCII, '?');
```
Comment (obelisk) [inline osquery/tables/system/darwin/mdls.mm]: Please avoid c style casts

File: osquery/tables/system/darwin/mdls.mm [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    }
+    if (value == nullptr) {
+      continue;
+    }
+    Row r;
+
+    CFStringRef valuetype = CFCopyTypeIDDescription(CFGetTypeID(value));
+    rvalue = getStringOfValue(value);
+
+    r["path"] = stringFromCFString((CFStringRef)tr);
```
Comment (obelisk) [inline osquery/tables/system/darwin/mdls.mm]: Please avoid c style casts

File: osquery/tables/system/darwin/mdls.mm [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+    r["key"] = stringFromCFString((CFStringRef)key);
```
Comment (obelisk) [inline osquery/tables/system/darwin/mdls.mm]: Please avoid c style casts

File: osquery/tables/system/darwin/mdls.mm [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+    r["value"] = rvalue;
+    r["valuetype"] = stringFromCFString((CFStringRef)valuetype);
```
Comment (obelisk) [inline osquery/tables/system/darwin/mdls.mm]: Please avoid c style casts

File: osquery/tables/system/darwin/mdls.mm [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+  CFStringRef cs = CFStringCreateWithCString(
+      kCFAllocatorDefault, path.c_str(), kCFStringEncodingASCII);
+  MDItemRef mdi = MDItemCreate(kCFAllocatorDefault, cs);
+  CFTypeRef tr = MDItemCopyAttribute(mdi, CFSTR("kMDItemPath"));
+  CFArrayRef al = MDItemCopyAttributeNames(mdi);
+  CFDictionaryRef d = MDItemCopyAttributes(mdi, al);
+  for (int j = 0; j < CFArrayGetCount(al); ++j) {
+    CFTypeRef key = CFArrayGetValueAtIndex(al, j);
+    CFTypeRef value = CFDictionaryGetValue(d, key);
+    std::string rvalue = "null";
```
Comment (obelisk) [inline osquery/tables/system/darwin/mdls.mm]: Nit, you can used braced initialization here.

**Comment by drakearonhalt** [comment]
Fixing the casts, sample output below. Note `path` is a required field in the query. 

Sample output: 


```+-----------+------------------------------------+------------------------------------------------------------------+
| path      | key                                | value                                                            |
+-----------+------------------------------------+------------------------------------------------------------------+
| /bin/date | kMDItemContentTypeTree             | public.item,public.executable,public.data,public.unix-executable |
| /bin/date | kMDItemSupportFileType             | MDSystemFile                                                     |
| /bin/date | kMDItemContentType                 | public.unix-executable                                           |
| /bin/date | kMDItemPhysicalSize                | 12288                                                            |
| /bin/date | kMDItemDisplayName                 | date                                                             |
| /bin/date | kMDItemKind                        | Terminal Document                                                |
| /bin/date | kMDItemExecutableArchitectures     | x86_64                                                           |
| /bin/date | kMDItemDateAdded                   | 1523543790                                                       |
| /bin/date | kMDItemContentModificationDate     | 1522209793                                                       |
| /bin/date | kMDItemContentCreationDate         | 1522209793                                                       |
| /bin/date | kMDItemLogicalSize                 | 28608                                                            |
| /bin/date | kMDItemDateAdded_Ranking           | 1523491200                                                       |
| /bin/date | kMDItemContentCreationDate_Ranking | 1522195200                                                       |
| /bin/date | kMDItemInterestingDate_Ranking     | 1522195200                                                       |
| /bin/date | kMDItemFSName                      | date                                                             |
| /bin/date | kMDItemFSSize                      | 28608                                                            |
| /bin/date | kMDItemFSCreationDate              | 1522209793                                                       |
| /bin/date | kMDItemFSContentChangeDate         | 1522209793                                                       |
| /bin/date | kMDItemFSOwnerUserID               | 0                                                                |
| /bin/date | kMDItemFSOwnerGroupID              | 0                                                                |
| /bin/date | kMDItemFSInvisible                 | false                                                            |
| /bin/date | kMDItemFSTypeCode                  | 0                                                                |
| /bin/date | kMDItemFSCreatorCode               | 0                                                                |
| /bin/date | kMDItemFSFinderFlags               | 0                                                                |
| /bin/date | kMDItemFSIsExtensionHidden         | false                                                            |
| /bin/date | kMDItemFSLabel                     | 0                                                                |
+-----------+------------------------------------+------------------------------------------------------------------+
```

**Comment by drakearonhalt** [comment]
Still working on the leaks but all the C-style casts should be removed.

**Comment by drakearonhalt** [comment]
@obelisk I believe I fixed all the requested changes. leaks test reports 0 leaks, all C casts have been removed. Let me know if I missed anything.

File: osquery/tables/system/darwin/mdls.mm [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+    auto unix_time = CFDateGetAbsoluteTime(static_cast<CFDateRef>(value)) +
+                     kCFAbsoluteTimeIntervalSince1970;
+    rvalue = INTEGER(std::llround(unix_time));
+  } else if (CFGetTypeID(value) == CFArrayGetTypeID()) {
+    for (int i = 0; i < CFArrayGetCount(static_cast<CFArrayRef>(value)); ++i) {
+      CFTypeRef b = CFArrayGetValueAtIndex(static_cast<CFArrayRef>(value), i);
+      if (i == 0) {
+        rvalue = getStringOfValue(b);
+      } else {
+        // beware of recursion....
```
Comment (obelisk) [inline osquery/tables/system/darwin/mdls.mm]: Should a max recursion depth be implemented here?

File: osquery/tables/system/darwin/mdls.mm [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+    //Do not release key or value, they are released when the dict is released
+    CFTypeRef key = CFArrayGetValueAtIndex(al, j);
+    CFTypeRef value = CFDictionaryGetValue(d, key);
+    std::string rvalue{"null"};
+
+    if (tr == nullptr || key == nullptr || value == nullptr) {
+      continue;
+    }
+
+    Row r;
```
Comment (obelisk) [inline osquery/tables/system/darwin/mdls.mm]: Move closer to usage

File: specs/darwin/mdls.table [hunk 14]
```diff
@@ -0,0 +1,13 @@
+table_name("mdls")
+description("Query file metadata in the Spotlight database.")
+schema([
+    Column("path", TEXT, "Path of the file", required=True),
+    Column("key", TEXT, "Name of the metadata key"),
+    Column("value", TEXT, "Value stored in the metadata key"),
+    Column("valuetype", TEXT, "CoreFoundation Type of data stored in value", hidden=True),
```
Comment (obelisk) [inline specs/darwin/mdls.table]: Nit. type

**Comment by jordan-rash** [comment]
Is there something I can help do to move this along?  Love this capability and look forward to using it

**Comment by FritzX6** [comment]
What's the status on implementation of this table? Is it dead in the water or still on the roadmap?

**Comment by drakearonhalt** [comment]
I made all the requested changes a couple months ago. It's based on the old master branch though might need some minor updates to merge into the current experimental branch.

**Comment by directionless** [comment]
There's some conversation that this `mdls` can list a bunch of data. So we may want to blacklist some fields. @FritzX6 any opinions?

**Comment by FritzX6** [comment]
We might want to consider the privacy implications of the ItemDescription field (which will grab iMessage transcripts) and some of the addressee metadata (which would return phone numbers of individuals you were messaging).

Example output below:
```
➜  ~ mdls '/Users/fritz-imac/Library/Messages/Archive/2019-06-18/Chat with <202a><202d>+1 (203) 313-2253<202c><202c> et al on 2019-06-18 at 10.22.32.ichat' 
_kMDItemDisplayNameWithExtensions  = "Chat with +1 (203) 313-2253 et al on 2019-06-18 at 10.22.32.ichat"
kMDItemAuthorAddresses             = (
    "e:"
)
kMDItemAuthors                     = (
    "Fritz Ifert-Miller",
    "\U202a\U202d+1 (203) 313-2253\U202c\U202c",
    "\U202a\U202d+1 (860) 733-2618\U202c\U202c",
    "\U202a\U202d+1 (203) 490-4876\U202c\U202c"
)
kMDItemContentCreationDate         = 2019-06-18 13:58:28 +0000
kMDItemContentCreationDate_Ranking = 2019-06-18 00:00:00 +0000
kMDItemContentModificationDate     = 2019-06-18 18:29:47 +0000
kMDItemContentType                 = "com.apple.ichat.transcript"
kMDItemCoverage                    = "chat809819206630457684"
kMDItemDateAdded                   = 2019-06-18 18:29:58 +0000
kMDItemDateAdded_Ranking           = 2019-06-18 00:00:00 +0000
kMDItemDeliveryType                = "SMS"
kMDItemDescription                 = "Hey you guys going to QueryCon this week? Absolutely, I'm giving a talk this year! Drinks 🍻 afterwards? You know it!"
kMDItemDisplayName                 = "Chat with +1 (203) 313-2253 et al on 2019-06-18 at 10.22.32.ichat”
kMDItemDurationSeconds             = 16278
kMDItemFSContentChangeDate         = 2019-06-18 18:29:47 +0000
kMDItemFSCreationDate              = 2019-06-18 14:22:32 +0000
kMDItemFSName                      = "Chat with +1 (203) 313-2253 et al on 2019-06-18 at 10.22.32.ichat"
kMDItemInstantMessageAddresses     = (
    "e:",
    12033132253,
    18607332618,
    12034904876
)
kMDItemInterestingDate_Ranking     = 2019-06-18 00:00:00 +0000
kMDItemIsApplicationManaged        = 1
```

However, with the introduction of Mojave and Catalina the `~/Library/Messages/Archive/%` directories are protected under TCC so it may be a non-issue.

File: osquery/tables/system/darwin/mdls.mm [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+  std::string rvalue;
+
+  if (CFGetTypeID(value) == CFNumberGetTypeID()) {
+    rvalue = stringFromCFNumber(static_cast<CFDataRef>(value));
+  } else if (CFGetTypeID(value) == CFDateGetTypeID()) {
+    auto unix_time = CFDateGetAbsoluteTime(static_cast<CFDateRef>(value)) +
+                     kCFAbsoluteTimeIntervalSince1970;
+    rvalue = INTEGER(std::llround(unix_time));
+  } else if (CFGetTypeID(value) == CFArrayGetTypeID()) {
+    for (int i = 0; i < CFArrayGetCount(static_cast<CFArrayRef>(value)); ++i) {
```
Comment (obelisk) [inline osquery/tables/system/darwin/mdls.mm]: nit auto

File: osquery/tables/system/darwin/mdls.mm (hunk 12 above)
Comment (obelisk) [reply osquery/tables/system/darwin/mdls.mm]: This has been addressed.

File: osquery/tables/system/darwin/mdls.mm [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+      }
+    }
+  } else if (CFGetTypeID(value) == CFBooleanGetTypeID()) {
+    if (CFBooleanGetValue(static_cast<CFBooleanRef>(value))) {
+      rvalue = "true";
+    } else {
+      rvalue = "false";
+    }
+  } else if (CFGetTypeID(value) == CFStringGetTypeID()) {
+    // might be able to just stringFromCFString after #4778
```
Comment (obelisk) [inline osquery/tables/system/darwin/mdls.mm]: Is this true now? Should it be used?

**Comment by drakearonhalt** [comment]
profile.py seems to have a minor issue that throws a TypeError when running the leaks test. Seems to be an easy fix though, leaks test results below. Same result running as sudo. 

```
Analyzing leaks in query: select * from mdls where path = '/Users/drake/Desktop/test.txt';
  definitely: 0 leaks for 0 total leaked bytes.
```

**Comment by drakearonhalt** [comment]
Ran a more slightly more comprehensive leaks test at @obelisk suggestion. 
```
Analyzing leaks in query: select * from mdls where path LIKE '/Users/drake/Downloads/%%';
  definitely: 0 leaks for 0 total leaked bytes.
```
```
./build/osquery/osqueryi "select count(*) from file where path LIKE '/Users/drake/Downloads/%%';"
+----------+
| count(*) |
+----------+
| 1105     |
+----------+

File: osquery/tables/system/darwin/mdls.mm [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+  } else {
+    rvalue = "null";
+  }
+
+  return rvalue;
+}
+void genResults(const std::string& path, QueryData& results) {
+  CFStringRef cs = CFStringCreateWithCString(
+      kCFAllocatorDefault, path.c_str(), kCFStringEncodingASCII);
+  MDItemRef mdi = MDItemCreate(kCFAllocatorDefault, cs);
```
Comment (zwass) [inline osquery/tables/system/darwin/mdls.mm]: Does this need to be released? Not clear based on the docs: https://developer.apple.com/documentation/coreservices/1426917-mditemcreate?language=objc

File: osquery/tables/system/darwin/mdls.mm [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+    // Do not release key or value, they are released when the dict is released
+    CFTypeRef key = CFArrayGetValueAtIndex(al, j);
+    CFTypeRef value = CFDictionaryGetValue(d, key);
+    std::string rvalue{"null"};
+
+    if (tr == nullptr || key == nullptr || value == nullptr) {
+      continue;
+    }
+
+    CFStringRef valuetype = CFCopyTypeIDDescription(CFGetTypeID(value));
```
Comment (zwass) [inline osquery/tables/system/darwin/mdls.mm]: Does this need to be released? I think so based on the docs: https://developer.apple.com/documentation/corefoundation/1521220-cfcopytypeiddescription?language=objc

Comment (drakearonhalt) [reply osquery/tables/system/darwin/mdls.mm]: `valuetype` gets released further down on line 89, after it's used. The code could be refactored a bit to make that clearer?

File: osquery/tables/system/darwin/mdls.mm (hunk 17 above)
Comment (drakearonhalt) [reply osquery/tables/system/darwin/mdls.mm]: I don't see any reason why we shouldn't release it.

File: osquery/tables/system/darwin/mdls.mm (hunk 18 above)
Comment (zwass) [reply osquery/tables/system/darwin/mdls.mm]: Ah yes I see that now. Thank you.

**Comment by drakearonhalt** [comment]
I didn't know `scope_guard` existed.  I don't have a problem with adding it if it makes the code easier to read. 

`auto const guard = scope_guard::create( [valuetype]() { CFRelease(valuetype); });`

File: osquery/tables/system/darwin/mdls.mm [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+  } else {
+    rvalue = "null";
+  }
+
+  return rvalue;
+}
+void genResults(const std::string& path, QueryData& results) {
+  CFStringRef cs = CFStringCreateWithCString(
+      kCFAllocatorDefault, path.c_str(), kCFStringEncodingASCII);
+  MDItemRef mdi = MDItemCreate(kCFAllocatorDefault, cs);
```
Comment (zwass) [inline osquery/tables/system/darwin/mdls.mm]: Check for nullptr?

File: osquery/tables/system/darwin/mdls.mm [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+  CFTypeRef tr = MDItemCopyAttribute(mdi, CFSTR("kMDItemPath"));
```
Comment (zwass) [inline osquery/tables/system/darwin/mdls.mm]: Check for nullptr?

File: osquery/tables/system/darwin/mdls.mm [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+  CFArrayRef al = MDItemCopyAttributeNames(mdi);
```
Comment (zwass) [inline osquery/tables/system/darwin/mdls.mm]: Check for nullptr?

File: osquery/tables/system/darwin/mdls.mm [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+  CFDictionaryRef d = MDItemCopyAttributes(mdi, al);
```
Comment (zwass) [inline osquery/tables/system/darwin/mdls.mm]: Check for nullptr?

File: osquery/tables/system/darwin/mdls.mm [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+  CFRelease(mdi);
+  for (int j = 0; j < CFArrayGetCount(al); ++j) {
+    // Do not release key or value, they are released when the dict is released
+    CFTypeRef key = CFArrayGetValueAtIndex(al, j);
+    CFTypeRef value = CFDictionaryGetValue(d, key);
+    std::string rvalue{"null"};
+
+    if (tr == nullptr || key == nullptr || value == nullptr) {
```
Comment (zwass) [inline osquery/tables/system/darwin/mdls.mm]: Can remove the check for `tr == nullptr` here if we check for it above.

File: osquery/tables/system/darwin/mdls.mm [hunk 24]
```diff
@@ -0,0 +1,130 @@
...
+      continue;
+    }
+
+    CFStringRef valuetype = CFCopyTypeIDDescription(CFGetTypeID(value));
```
Comment (zwass) [inline osquery/tables/system/darwin/mdls.mm]: Check for nullptr?

File: osquery/tables/system/darwin/mdls.mm [hunk 25]
```diff
@@ -60,25 +61,56 @@
 
   return rvalue;
 }
+
 void genResults(const std::string& path, QueryData& results) {
   CFStringRef cs = CFStringCreateWithCString(
       kCFAllocatorDefault, path.c_str(), kCFStringEncodingASCII);
+  auto const cs_guard = scope_guard::create([cs]() { CFRelease(cs); });
```
Comment (zwass) [inline osquery/tables/system/darwin/mdls.mm]: These statements need to be swapped as `CFRelease` will be called with `nullptr` (and crash the program) if the allocation function returns null. 

Is it possible to use `cs` by reference in the lambda to save an allocation?

Same concerns below.

**Comment by obelisk** [comment]
@zwass Are you good with this as it stands now?

**Pull request merged**